WalletWasabi.Tests/UnitTests/EventSourcing/InMemoryEventRepositoryTests.cs

```diff
@@ -0,0 +1,438 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using WalletWasabi.EventSourcing;
+using WalletWasabi.Exceptions;
+using WalletWasabi.Interfaces.EventSourcing;
+using WalletWasabi.Tests.UnitTests.EventSourcing.TestDomain;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace WalletWasabi.Tests.UnitTests.EventSourcing
+{
+	public class InMemoryEventRepositoryTests : IDisposable
+	{
+		private const string TestRoundAggregate = "TestRoundAggregate";
+		private readonly TimeSpan _semaphoreWaitTimeout = TimeSpan.FromSeconds(20);
+
+		public InMemoryEventRepositoryTests(ITestOutputHelper output)
+		{
+			TestEventRepository = new TestInMemoryEventRepository(output);
+			EventRepository = TestEventRepository;
+		}
+
+		private IEventRepository EventRepository { get; init; }
+		private TestInMemoryEventRepository TestEventRepository { get; init; }
+
+		[Fact]
+		public async Task AppendEvents_Zero_Async()
+		{
+			// Arrange
+			var events = Array.Empty<WrappedEvent>();
+
+			// Act
+			await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events);
+
+			// Assert
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "1"))
+				.SequenceEqual(events));
+			Assert.True((await EventRepository.ListAggregateIdsAsync(nameof(TestRoundAggregate)))
+				.SequenceEqual(Array.Empty<string>()));
+		}
+
+		[Fact]
+		public async Task AppendEvents_One_Async()
+		{
+			// Arrange
+			var events = new[]
+			{
+				new WrappedEvent(1),
+			};
+
+			// Act
+			await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events);
+
+			// Assert
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "1"))
+				.SequenceEqual(events));
+			Assert.True((await EventRepository.ListAggregateIdsAsync(nameof(TestRoundAggregate)))
+				.SequenceEqual(new[] { "1" }));
+		}
+
+		[Fact]
+		public async Task AppendEvents_Two_Async()
+		{
+			// Arrange
+			var events = new[]
+			{
+				new WrappedEvent(1),
+				new WrappedEvent(2),
+			};
+
+			// Act
+			await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events);
+
+			// Assert
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "1"))
+				.SequenceEqual(events));
+			Assert.True((await EventRepository.ListAggregateIdsAsync(nameof(TestRoundAggregate)))
+				.SequenceEqual(new[] { "1" }));
+		}
+
+		[Fact]
+		public async Task AppendEvents_NegativeSequenceId_Async()
+		{
+			// Arrange
+			var events = new[]
+			{
+				new WrappedEvent(-1)
+			};
+
+			// Act
+			async Task ActionAsync()
+			{
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events);
```

Comment:
local functions can contain calls with `ConfigureAwait(false)` and that should help with test performance.

```suggestion
				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events).ConfigureAwait(false);
```

Reply:
not in tests. Tests are not reusable library.

Reply:
IMO there is not performance problem here. The tests are just in memory and super fast. in general .ConfigureAwait(false) is not required in tests

Reply:
the test failed most probably because the server which run tests on github doesn't have allocated much cpu time so it takes a lot of real time to get enough cpu time. so it is only about ratio of real time to cpu time. I increased the timeout and it works.

Reply:
xUnit has a synchronization context so when you use `ConfigureAwait(true)` you may wait longer than if you just write `ConfigureAwait(false)` at mentioned places. 

Anyway, it's also a convention in test code to use `ConfigureAwait(true)` directly in test methods and `ConfigureAwait(false)` in the remaining cases.

Some notes are here: https://github.com/zkSNACKs/WalletWasabi/issues/4147

Reply:
No it doesn't make sense. ConfigureAwait(false) is not necessary in tests. But whatever I will do it. This is so stupid.